### ISSUE TITLE
reward 无法上升

### ISSUE THREAD

**Issue by SXUSongJH** [XueZeyue/DanceGRPO#45]
你好，感谢作者出色的工作。
我是强化学习的一名初学者，我目前在使用 GRPO 调优音频生成模型（类似于SD3），我在训练过程中发现我的 reward 无法上升，且训练了 1500 steps 后容易出现 “策略崩塌”，reward 急剧下降，样本生成出来的是噪音。我的训练参数如下：
- num_gpus = 8
- sample_batch_size = 1
- num_generations = 16
- train_batch_size = 4
- gradient_accumulation_steps = sample_steps = 20
- learning_rate = 1e-6
- clip_range = 1e-2
- cfg_scale=15.0

即每个batch，每张GPU上都会分配到 1 个prompt (各种卡promt不一样)，然后用这个prompt生成16个样本，计算 reward 以及 advantage，然后我再将这 16 个样本打包成 batch_size 为 4 的batches 用于训练，计算 GRPO loss，再所有卡汇总loss，更新梯度。然而我的训练结果如下图：

<img width="1354" height="590" alt="Image" src="https://github.com/user-attachments/assets/04bc312e-a6b5-4044-b232-e96c983c9ba1" />

想请教一下作者我的训练过程是否有问题？该如何改善训练效果？我目前怀疑是不是 learning_rate 或者是 batch_size 太小了？

**Comment by XueZeyue** [comment]
My debug process:
1. Print the probability ratio, reward, and advantage for each sample; the ratio should be 1.0 before the gradient update, and you can verify the advantage on your own.
2. The gradient accumulation should follow the sample dimension, which means, suppose you use 20 steps, the gradient accumulation should be accumulate_samples*20
3. Based on my experience, the learning rate should be set to between 5e-6 and 2e-5, setting the lr to 1e-6 always leads to training failure in my settings.
4. Maybe the batchsize is not enough, you can follow my setting of flux_8gpus
5. More importantly, if you enable cfg, the gradient accumulation should be set to a large number, based on my experience, I always set it to be num_generations*20, which means you just update the gradient for one time in each rollout.

**Comment by SXUSongJH** [comment]
感谢作者的回复！
我检查了训练过程中的 ratio，发现 ratio 绝大多数是 0.99xx, 0.98xx 这样接近 1.0 的数值，但偶尔会出现 0.5 以下的极端情况。

请问这表明我的训练代码是有问题的吗？另外，能详细解释以下你提到的第5点，为什么在开启 cfg 时，需要设置更大的 gradient accumulation?

**Comment by XueZeyue** [comment]
There might be some bugs. Can you try to set the rollout inference batchsize to one?

CFG sometimes will lead to training instability; a larger gradient accumulation helps stabilize it. And, don't forget to backpropagate both on the outputs of the prompt and the null prompt

**Comment by SXUSongJH** [comment]
@sduzzx857 
不正常，如果你是参考的 DanceGRPO 的代码，在 update gredient 之前计算 ratio，那 ratio 必须为1.0，因为计算log prob 的模型的参数是一致的，如果不是 1.0 ，说明训练代码中可能哪里引入了误差。

**Comment by darkliang** [comment]
我这边第一步的ratio是0.9999x，是不是bf16的误差呢

**Comment by SXUSongJH** [comment]
> 我这边第一步的ratio是0.9999x，是不是bf16的误差呢

估计是的，你可以把 gradient accumulation 开大一点，多看几步没有更新梯段前计算的 ratio

**Comment by cfeng16** [comment]
你们的ratio是每一个iteration里的第一个step是1.0还是iteration里的所有steps都有1.0
我看到我的输出是所有steps都是1.0，但理论上来说应该只有第一个step是1.0?

**Comment by cfeng16** [comment]
为什么是gradient_acc_step== num_generations？
gradient_acc_step内确实没有梯度更新，所有step应该也是1.0
但是gradient_acc_step应该和num_generations没有什么关系？

**Comment by SXUSongJH** [comment]
> 为什么是gradient_acc_step== num_generations？ gradient_acc_step内确实没有梯度更新，所有step应该也是1.0 但是gradient_acc_step应该和num_generations没有什么关系？

你可以看一下作者回复的第5点

**Comment by XueZeyue** [comment]
> ok，我用的是video的dancegrpo，我的gradient accumulation steps=4 bestofn=8 num_generations=24 train_batch size=1 但即使是在一个iteration内（因为bestofn是8，所以要走完8个steps?)，超过了gradient accumulation steps，ratio仍然是1，这是不是说明model没有更新？主要是我自己换了reward model （reward比较小，大部分都是0.0x)，不知道是不是这里出了问题。。。

video的gradient accumulation step我i2v开的8，t2v开的4，都是bestofn=8，有兴趣可以关注下昨天混元放的MixGRPO，我觉得会对视频加速有帮助

**Comment by SXUSongJH** [comment]
> > ok，我用的是video的dancegrpo，我的gradient accumulation steps=4 bestofn=8 num_generations=24 train_batch size=1 但即使是在一个iteration内（因为bestofn是8，所以要走完8个steps?)，超过了gradient accumulation steps，ratio仍然是1，这是不是说明model没有更新？主要是我自己换了reward model （reward比较小，大部分都是0.0x)，不知道是不是这里出了问题。。。
> 
> video的gradient accumulation step我i2v开的8，t2v开的4，都是bestofn=8，有兴趣可以关注下昨天混元放的MixGRPO，我觉得会对视频加速有帮助

请问大佬，我设置了 accmulation_steps = sample_steps，意味着在第个 minibatch 后就会更新一次梯度，但我发现在更新梯度后的 minibatch 中，计算出来的 ratio 仍然等于 1.0000，没有超出 1 的情况，这正常吗？感觉是不是参数压根没更新？

**Comment by XueZeyue** [comment]
> > > ok，我用的是video的dancegrpo，我的gradient accumulation steps=4 bestofn=8 num_generations=24 train_batch size=1 但即使是在一个iteration内（因为bestofn是8，所以要走完8个steps?)，超过了gradient accumulation steps，ratio仍然是1，这是不是说明model没有更新？主要是我自己换了reward model （reward比较小，大部分都是0.0x)，不知道是不是这里出了问题。。。
> > 
> > 
> > video的gradient accumulation step我i2v开的8，t2v开的4，都是bestofn=8，有兴趣可以关注下昨天混元放的MixGRPO，我觉得会对视频加速有帮助
> 
> 请问大佬，我设置了 accmulation_steps = sample_steps，意味着在第个 minibatch 后就会更新一次梯度，但我发现在更新梯度后的 minibatch 中，计算出来的 ratio 仍然等于 1.0000，没有超出 1 的情况，这正常吗？感觉是不是参数压根没更新？

看你开不开cfg，不开cfg的话ratio变动很小，开了cfg的话变化会大不少
但是最终reward结果应该都还可以，我觉得是因为diffusion生成的是一个连续概率分布，不是离散分布，连续分布的话很难造成类似AR模型那么大的扰动，我觉得这也是ddpo把clip那个概率ratio开成1e-4的原因，这个东西在llm一般是0.1或者0.3

**Comment by XueZeyue** [comment]
昨晚想到了另一个事儿，我这段代码是fork DDPO的，当时做的时候没仔细想这里

      # mean along all but batch dimension
      log_prob = log_prob.mean(dim=tuple(range(1, log_prob.ndim)))

但现在想来，这个做法更接近于GSPO，即概率是在图像维度计算的，LLM那边一开始是在token维度，现在也是在sequence维度了，我感觉是因为我这里平均了，导致ratio变化不够大
